sensing/autoware_pointcloud_preprocessor/src/concatenate_data/combine_cloud_handler.cpp

```diff
@@ -0,0 +1,301 @@
+CombineCloudHandler::combinePointClouds(
+  std::unordered_map<std::string, sensor_msgs::msg::PointCloud2::SharedPtr> & topic_to_cloud_map)
+{
+  sensor_msgs::msg::PointCloud2::SharedPtr concatenate_cloud_ptr = nullptr;
+  std::unordered_map<std::string, sensor_msgs::msg::PointCloud2::SharedPtr>
+    topic_to_transformed_cloud_map;
+  std::unordered_map<std::string, double> topic_to_original_stamp_map;
+
+  std::vector<rclcpp::Time> pc_stamps;
+  for (const auto & pair : topic_to_cloud_map) {
+    pc_stamps.push_back(rclcpp::Time(pair.second->header.stamp));
+  }
+  std::sort(pc_stamps.begin(), pc_stamps.end(), std::greater<rclcpp::Time>());
+  const auto oldest_stamp = pc_stamps.back();
+
+  std::unordered_map<rclcpp::Time, Eigen::Matrix4f, RclcppTimeHash_> transform_memo;
+
+  for (const auto & pair : topic_to_cloud_map) {
+    std::string topic = pair.first;
+    sensor_msgs::msg::PointCloud2::SharedPtr cloud = pair.second;
+
+    auto transformed_cloud_ptr = std::make_shared<sensor_msgs::msg::PointCloud2>();
+    if (output_frame_ != cloud->header.frame_id) {
+      if (!pcl_ros::transformPointCloud(
+            output_frame_, *cloud, *transformed_cloud_ptr, tf_buffer_)) {
+        RCLCPP_ERROR(
+          node_->get_logger(),
+          "Transform pointcloud from %s to %s failed, Please check the defined output frame.",
+          cloud->header.frame_id.c_str(), output_frame_.c_str());
+        transformed_cloud_ptr = cloud;
+      }
+    } else {
+      transformed_cloud_ptr = cloud;
+    }
+
+    topic_to_original_stamp_map[topic] = rclcpp::Time(cloud->header.stamp).seconds();
+
+    auto transformed_delay_compensated_cloud_ptr =
+      std::make_shared<sensor_msgs::msg::PointCloud2>();
+
+    if (is_motion_compensated_) {
+      Eigen::Matrix4f adjust_to_old_data_transform = Eigen::Matrix4f::Identity();
+      rclcpp::Time current_cloud_stamp = rclcpp::Time(cloud->header.stamp);
+      for (const auto & stamp : pc_stamps) {
+        if (stamp >= current_cloud_stamp) continue;
+
+        Eigen::Matrix4f new_to_old_transform;
+        if (transform_memo.find(stamp) != transform_memo.end()) {
+          new_to_old_transform = transform_memo[stamp];
+        } else {
+          new_to_old_transform =
+            computeTransformToAdjustForOldTimestamp(stamp, current_cloud_stamp);
+          transform_memo[stamp] = new_to_old_transform;
+        }
+        adjust_to_old_data_transform = new_to_old_transform * adjust_to_old_data_transform;
+        current_cloud_stamp = stamp;
+      }
+      pcl_ros::transformPointCloud(
+        adjust_to_old_data_transform, *transformed_cloud_ptr,
+        *transformed_delay_compensated_cloud_ptr);
+
+    } else {
+      transformed_delay_compensated_cloud_ptr = transformed_cloud_ptr;
+    }
+
+    // concatenate
+    if (concatenate_cloud_ptr == nullptr) {
+      concatenate_cloud_ptr =
+        std::make_shared<sensor_msgs::msg::PointCloud2>(*transformed_delay_compensated_cloud_ptr);
```

Comment:
@mojomex 
Thanks for the suggestion!!!
fixed thme in https://github.com/autowarefoundation/autoware.universe/pull/8300/commits/49b54d4d49865b7745791d4204f2a1593c1122a2

Reply:
Sorry, this is not quite what I meant: with your new implementation, the size of the additional single pointlcoud is reserved but not the size of all poinclouds combined.

What I wanted to see is
1. initialize `concatenate_cloud_ptr` with an empty pointcloud (*1) before the for loop starts
2. reserve `data` such that it's (pseudocode) `point_step * sum([cloud.data.size() for cloud in cloud_map.values()]` (so if you have clouds with sizes 1, 2 and 3 bytes, it would reserve 6 bytes **at once** before the for loop starts
3. then concatenate the clouds to `concatenate_cloud_ptr` in the loop

The manual initialization of all fields should not be necessary (please revert to how it was before, just make_shared).

*1: when calling `pcl::concatenatePointCloud` it will sadly throw away the empty initial pointcloud... Unless we want to do something sketchy (or implement concatenation ourselves) we can't proceed with this comment. Feel free to revert and resolve this conversation, but **please add a comment above the for loop** saying that this could be made faster by reserving space first.

Reply:
@mojomex 
Thanks, understand now.
Fixed in the commit https://github.com/autowarefoundation/autoware.universe/pull/8300/commits/782228fb9b1e97b5f066647c37ae357324ad2b71